pkg/cmd/server/api/validation/master.go

```diff
@@ -0,0 +1,169 @@
+package validation
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+
+	errs "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+
+	"github.com/openshift/origin/pkg/cmd/server/api"
+)
+
+func ValidateMasterConfig(config *api.MasterConfig) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	if config.AssetConfig != nil {
+		allErrs = append(allErrs, ValidateAssetConfig(config.AssetConfig).Prefix("assetConfig")...)
+		colocated := config.AssetConfig.ServingInfo.BindAddress == config.ServingInfo.BindAddress
+		if colocated {
+			publicURL, _ := url.Parse(config.AssetConfig.PublicURL)
+			if publicURL.Path == "/" {
+				allErrs = append(allErrs, errs.NewFieldInvalid("assetConfig.publicURL", config.AssetConfig.PublicURL, "path can not be / when colocated with master API"))
+			}
+		}
+
+		if config.OAuthConfig != nil && config.OAuthConfig.AssetPublicURL != config.AssetConfig.PublicURL {
+			allErrs = append(allErrs,
+				errs.NewFieldInvalid("assetConfig.publicURL", config.AssetConfig.PublicURL, "must match oauthConfig.assetPublicURL"),
+				errs.NewFieldInvalid("oauthConfig.assetPublicURL", config.OAuthConfig.AssetPublicURL, "must match assetConfig.publicURL"),
+			)
+		}
+
+		// TODO warn when the CORS list does not include the assetConfig.publicURL host:port
+		// only warn cause they could handle CORS headers themselves in a proxy
+	}
+
+	if config.DNSConfig != nil {
+		allErrs = append(allErrs, ValidateHostPort(config.DNSConfig.BindAddress, "bindAddress").Prefix("dnsConfig")...)
+	}
+
+	allErrs = append(allErrs, ValidateEtcdConnectionInfo(config.EtcdClientInfo).Prefix("etcdClientInfo")...)
```

Comment:
Given both etcdConnectionInformation and etcdConfig, it is possible to know if they match.  Starting in process and then choosing to expose through a proxy is probably insane.